src/librustc_mir/interpret/validity.rs

```diff
@@ -475,15 +475,25 @@ impl<'rt, 'mir, 'tcx, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, 'tcx, M
                 Ok(true)
             }
             ty::RawPtr(..) => {
-                // We are conservative with undef for integers, but try to
-                // actually enforce our current rules for raw pointers.
-                let place = try_validation!(
-                    self.ecx.ref_to_mplace(self.ecx.read_immediate(value)?),
-                    "undefined pointer",
-                    self.path
-                );
-                if place.layout.is_unsized() {
-                    self.check_wide_ptr_meta(place.meta, place.layout)?;
+                let imm = self.ecx.read_immediate(value)?;
+                let pointee_kind = &imm.layout.ty.builtin_deref(true).unwrap().ty.kind;
+                match pointee_kind {
+                    ty::Param(_) => {
+                        // Creating pointers to T is valid. We only reach this case for unused
+                        // associated consts.
```

Comment:
This change only affects *raw* pointers. And it anyway was only checking that the pointer itself is not `Undef` -- whatever the pointer points to is already irrelevant here. I don't see how this change makes any sense.

However, validity certainly assumes that the type is fully known. It makes little sense to call validity at all when there are still `ty::Param` around.

Reply:
I still have `ty::Param` around in cases for unused associated consts, unfortunately, because I don't have a use-site to pull a type from. When `ref_to_mplace` calls `layout_of`, because the type is unknown, the layout is unknown, and validation fails. This breaks code which like this:

https://github.com/rust-lang/rust/blob/8045865873f7cdbb864d0f66ef5ecb0d3ad847b2/src/test/ui/consts/const-eval/ice-generic-assoc-const.rs#L9-L10

I didn't realise that this was just checking for `Undef` of the pointer value itself. I got the impression the purpose of this check was to check that the pointee had a valid layout (ie, that it's okay to dereference the raw pointer), so I thought skipping this check made sense because it's not until the pointer is dereferenced that any issues could occur.

I see a lot of const errors are picked up during `const_eval_raw`, before validation even happens - does it make sense to just not run validation for unused associated consts, then?

Reply:
> ie, that it's okay to dereference the raw pointer

Raw pointers may dangle. So such a check would be incorrect.

> I still have ty::Param around in cases for unused associated consts, unfortunately, because I don't have a use-site to pull a type from. 

Such constant should not have validation run on them, anyway, We cannot validate them, after all, with some type information missing.

I get the feeling you are sticking values into compiler passes that are not prepared to handle those values. I don't think this approach can work. In other words, I think this (running const-eval on associated consts) is the wrong approach to fix https://github.com/rust-lang/rust/issues/69021.

Reply:
> I think this (running const-eval on associated consts) is the wrong approach to fix #69021

After the poking I've done here, I've come to realise the issue is a bit broader than just missing bitshifts. We only get lints for associated consts if `const_eval` reaches the const through some use of it. That is, all lints for associated consts are missed if the associated const is unused. For example, the out-of-range access here isn't picked up if `S::N` is left unused:  

```rust
trait Foo {
    const N: i32;
}

struct S;
impl Foo for S {
    const N: i32 = [10][1];
}

fn main() {
    //println!("{}", S::N);
}

```

As for running const-eval being the wrong approach, I'd argue that associated consts are suitable for const eval in much of the same way normal consts are: by definition, they, like normal consts, must be able to be evaluated at compile time, and they also can be complex enough that just linting the AST without doing at least _some_ form of const evaluation isn't enough.

> I get the feeling you are sticking values into compiler passes that are not prepared to handle those values.

Mmm, I agree with your assessment. However, for the reason above, I'm inclined towards making const-eval work (but perhaps shoehorning `const_eval_poly` into this is the wrong approach).

Reply:
> As for running const-eval being the wrong approach, I'd argue that associated consts are suitable for const eval in much of the same way normal consts are: by definition, they, like normal consts, must be able to be evaluated at compile time, and they also can be complex enough that just linting the AST without doing at least some form of const evaluation isn't enough.

This only makes sense for *monomorphized* associated consts. It makes no sense when there are still unknown generics around, like what you are seeing here.

Associated consts with missing generics are much more like const *functions* than consts. And for const *functions*, we don't run const-eval either. And indeed the entire concept of executing such a constant/function is meaningless as some inputs are missing!

---

I really wonder why you are so focused on const_eval. Most of these lints don't come from const_eval. They come from const_prop. That's how the lint is triggered in run-time (non-const) functions, [such as this](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=b40c1bd271f0203e681b96373c4fcee9).

Here is the code triggering the lint:
https://github.com/rust-lang/rust/blob/8ab82b87af4f20b6c0a481e050517103d50263e9/src/librustc_mir/transform/const_prop.rs#L536-L554

const_prop is an optimization pass that *expects* to run on polymorphic functions with partial information, unlike const_eval which expects to run on fully monomorphic code. Thus I think the right fix here is to run const_prop on associated consts. That is also why [in this comment](https://github.com/rust-lang/rust/issues/69021#issuecomment-590410796) I talked about the relevant parts of const_prop. I wonder if there was any specific reason you went for const_eval?

Reply:
> as some inputs are missing!

I actually love that

```rust
impl<T> Nullable for *const T { 
     const NULL: Self = core::ptr::null::<T>(); 
```

is evaluable polymorphically. Not sure if this is relevant for @davidtwco.

If a constant (associated or not) can be evaluated polymorphically, then we don't need to run it again. We may experiment with this in `const_eval_raw`.

*back to topic:*

I think we should do something where associated constants and regular constants get handled the same way by `UnusedBrokenConst`. Not sure what that is, but the current PR is heading in that direction although maybe not on the path that we want. The reason we even have `UnusedBrokenConst` instead of doing everything in `const_prop` is the statement at the entry function of `const_prop`: we're going to evaluate the constant anyway.

So... I'm not sure what I think is the best way forward, but in general doing `const_prop` is more powerful because it doesn't bail out early if there's an error.

Reply:
I feel associated consts are more like `(const) fn`: there might be inputs missing. Thus we need const_prop, which is prepared for that case.

Of course, the actual "propagation" part of const_prop makes little sense here. Maybe putting lint and transformation together was just a bad call? (Another weird effect of this is that some warnings only appear in release mode, [such as this one](https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=b40c1bd271f0203e681b96373c4fcee9).)

Reply:
> I wonder if there was any specific reason you went for const_eval?

I looked into how unused non-assoc constants were linted and found UnusedBrokenConst, which runs `const_eval_poly`. I was hoping running `const_eval_poly` would handle all the cases where there weren't missing substitutions, and just bail out with TooGeneric in other cases:

https://github.com/rust-lang/rust/blob/285519d412ef9c65df3bcd2de2b1a3d6ca16a255/src/librustc/mir/interpret/queries.rs#L10-L13

On a second reading, and in light of everything you've both informed me here, I've realised I've focussed on the wrong part of that doc comment.

> Associated consts with missing generics are much more like const functions than consts. And for const functions, we don't run const-eval either. And indeed the entire concept of executing such a constant/function is meaningless as some inputs are missing!

> const_prop is an optimization pass that expects to run on polymorphic functions with partial information, unlike const_eval which expects to run on fully monomorphic code.

> in general doing const_prop is more powerful because it doesn't bail out early if there's an error

Ah, and that's why `const_eval_poly` was suitable for normal consts, but is not suitable for associated consts.

> Maybe putting lint and transformation together was just a bad call

I'm inclined to agree, and moreso when the transformation isn't always run. I'll pull the linting code into its own function as a very first step.